src/ICSharpCode.SharpZipLib/Zip/ZipFile.cs

```diff
@@ -3349,11 +3357,21 @@ private void DisposeInternal(bool disposing)
 				isDisposed_ = true;
 				entries_ = Empty.Array<ZipEntry>();
 
-				if (IsStreamOwner && (baseStream_ != null))
+				if (disposing)
 				{
-					lock (baseStream_)
+					try
 					{
-						baseStream_.Dispose();
+						this.baseStreamSemaphore.Wait();
```

Comment:
I did this the other day and wasn't sure about it, and it still feels off - if the idea of the existing lock is to prevent the stream being disposed while being read (rather than 'just' being disposed in between reads) then that needs to be maintained, but then you might also need to avoid disposing the semaphore while another thread is using it, which just makes things more complicated.

Still not sure if the lock is doing anything other than trying to save users from their own mistakes though (preventing an explicit attempt to dispose the archive while still using it) :-(

Reply:
Yeah, I'm not even sure that that is something we should try to manage. Tried looking up when the locking was added, but it lead me to [this wonderful commit](https://github.com/icsharpcode/SharpZipLib/commit/88e49b8a6acd09c647809351bf2679920582a429) with 124,411 additions and 10,927 deletions... Great.